### PR TITLE
[FX] Prototype Conv/BN fuser in FX

### PR DESCRIPTION
**Pull request by Chillee**
Some interesting stuff going on. All benchmarks are tested with both my implementation as well as the current quantized fuser.

For these benchmarks, things like using MKLDNN/FBGEMM make a big differene.

## Manual compilation (everything turned off)
In the small case, things look good
```
non-fused:  1.174886703491211
fused:  0.7494957447052002
```

However, for `torchvision.resnet18`, we see 
```
non-fused:  1.2272708415985107
fused:  3.7183213233947754
```

This is because Conv (no bias) -> Batch Norm is actually faster than Conv (bias) if you don't have any libraries...

## Nightly (CPU)
```
Toy
non-fused:  0.45807552337646484
fused:  0.34779977798461914

resnet18
non-fused:  0.14216232299804688
fused:  0.13438796997070312

resnet50
non-fused:  0.2999534606933594
fused:  0.29364800453186035

densenet161
non-fused:  0.6558926105499268
fused:  0.6190280914306641

inception_v3
non-fused:  1.2804391384124756
fused:  1.181272029876709
```
with MKLDNN. 

We see a small performance gain across the board, with more significant performance gains for smaller models.

## Nightly (CUDA)

```
M
non-fused:  1.2220964431762695
fused:  1.0833759307861328

resnet18
non-fused:  0.09721899032592773
fused:  0.09089207649230957

resnet50
non-fused:  0.2053072452545166
fused:  0.19138741493225098

densenet161
non-fused:  0.6830024719238281
fused:  0.660109281539917
```

### CODE REVIEW

File: torch/fx/experimental/fuser.py [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+    fused_conv = copy.deepcopy(conv)
+
+    fused_conv.weight, fused_conv.bias = \
+        fuse_conv_bn_weights(fused_conv.weight, fused_conv.bias,
+                             bn.running_mean, bn.running_var, bn.eps, bn.weight, bn.bias)
+
+    # fused_conv.bias = None
+    return fused_conv
+
+def fuse_conv_bn_weights(conv_w, conv_b, bn_rm, bn_rv, bn_eps, bn_w, bn_b):
```
Comment (vadimkantorov) [inline torch/fx/experimental/fuser.py:50]: I think this code exists in `torch.nn.utils.fusion.fuse_conv_bn_eval(conv, bn)` - would be good to use it directly?

File: torch/fx/experimental/fuser.py [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+                with model.graph.inserting_before(node.args[0]):
+                    new_node = model.graph.call_module(new_name, node.args[0].args)
+                node.replace_all_uses_with(new_node)
+                t_args = node.args[0]
+                node.args[0].replace_all_uses_with(new_node)
+                model.graph.erase_node(t_args)
+                model.graph.erase_node(node)
+    return GraphModule(model, model.graph)
+
+def fuse_conv_bn_eval(conv, bn):
```
Comment (vadimkantorov) [inline torch/fx/experimental/fuser.py:38]: This code already exists in `torch.nn.utils.fusion.fuse_conv_bn_eval(conv, bn)`. Is it unusable directly?

Comment (Chillee) [reply torch/fx/experimental/fuser.py:38]: No it's definitely usable haha - it's copy pasted from there. There were just some comparisons I wanted to make in different doing things while fusing.

**Comment by jamesr66a** [comment]
BTW probably a good idea to add a test:

```
diff --git a/test/test_fx_experimental.py b/test/test_fx_experimental.py
index 884e21a17dd..d10e55ac959 100644
--- a/test/test_fx_experimental.py
+++ b/test/test_fx_experimental.py
@@ -1,4 +1,5 @@
 import torch
+import unittest
 from typing import Dict
 from torch.fx.symbolic_trace import symbolic_trace
 from torch.fx.graph_module import GraphModule
@@ -15,8 +16,16 @@ from torch.fx.experimental.partitioner_utils import (
     Device,
     PartitionerConfig
 )
+from torch.fx.experimental.fuser import fuse
 from typing import Union, Callable
 
+try:
+    from torchvision.models import resnet18
+    HAS_TORCHVISION = True
+except ImportError:
+    HAS_TORCHVISION = False
+skipIfNoTorchVision = unittest.skipIf(not HAS_TORCHVISION, "no torchvision")
+
 
 def symbolic_trace_with_rewrite(root: Union[torch.nn.Module, Callable]) -> GraphModule:
     return GraphModule(
@@ -393,6 +402,19 @@ class TestFXExperimental(JitTestCase):
         # Confirm that the output is correct
         self.assertEqual(traced(3, 3), m(3, 3))
 
+    @skipIfNoTorchVision
+    def test_conv_bn_fusion(self):
+        rn18 = resnet18().eval()
+        traced = symbolic_trace(rn18)
+        fused = fuse(traced)
+
+        self.assertTrue(all(not isinstance(m, torch.nn.BatchNorm2d) for m in fused.modules()))
+
+        N, C, H, W = 20, 3, 224, 224
+        inp = torch.randn(N, C, H, W)
+
+        self.assertEqual(fused(inp), rn18(inp))
+
     def test_call_to_assert_with_msg(self):
         class M(torch.nn.Module):
             def forward(self, a, b):
```

File: torch/fx/experimental/fuser.py [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+            return False
+        if current_node.target not in modules:
+            return False
+        if type(modules[current_node.target]) is not pattern_node:
+            return False
+    return True
+
+
+def fuse(model):
+    patterns = [(torch.nn.Conv2d, torch.nn.BatchNorm2d), (torch.nn.Conv1d, torch.nn.BatchNorm1d), (torch.nn.Conv3d, torch.nn.BatchNorm3d)]
```
Comment (jamesr66a) [inline torch/fx/experimental/fuser.py:33]: nit: can we sort these numerically? :p

File: torch/fx/experimental/fuser.py [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+# Works for length 2 patterns with 2 modules
+def matches_module_pattern(pattern, node, modules):
+    if len(node.args) == 0:
+        return False
+    nodes = [node.args[0], node]
+    for pattern_node, current_node in zip(pattern, nodes):
+        if current_node.op != 'call_module':
+            return False
+        if not isinstance(current_node, torch.fx.Node):
+            return False
```
Comment (jamesr66a) [inline torch/fx/experimental/fuser.py:24]: If `current_node` isn't a `Node`, the check `current_node.op != 'call_module'` is going to throw, isn't it? This probably needs to go before the previous check. BTW mypy should catch this stuff if the fn is type annotated

File: torch/fx/experimental/fuser.py [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+        if current_node.target not in modules:
+            return False
+        if type(modules[current_node.target]) is not pattern_node:
```
Comment (jamesr66a) [inline torch/fx/experimental/fuser.py:27]: The name `pattern_node` is pretty misleading if it's a type. Probably want to name it something like `expected_module_type`

File: torch/fx/experimental/fuser.py [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+from torch.nn.utils.fusion import fuse_conv_bn_eval
+import torch
+
+# turn foo.bar -> ['foo', 'bar']
+def _parent_name(target):
+    r = target.rsplit('.', 1)
+    if len(r) == 1:
+        return '', r[0]
+    else:
+        return r[0], r[1]
```
Comment (jamesr66a) [inline torch/fx/experimental/fuser.py:13]: code golf

```
def _parent_name(target : str) -> Tuple[str, str]:
    *parent, name  = target.rsplit('.', 1)
    return parent if parent else '', name
```

File: torch/fx/experimental/fuser.py [hunk 7]
```diff
@@ -0,0 +1,45 @@
+from torch.fx import (
+    GraphModule,
+)
```
Comment (jamesr66a) [inline torch/fx/experimental/fuser.py:3]: I see there's a mypy error here. You can try 2 things to fix this:

1. `import torch.fx` and use `GraphModule` fully-qualified
2. add `__all__` to `torch/fx/__init__.py` and include all the exported names from that file. I think this is the "right" way to fix this

File: torch/fx/experimental/fuser.py [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+from torch.nn.utils.fusion import fuse_conv_bn_eval
+import torch
+
+# turn foo.bar -> ['foo', 'bar']
```
Comment (jamesr66a) [inline torch/fx/experimental/fuser.py:7]: Probably better to add a docstring to this function clarifying that this splits a qualname into its parent path and last atom. As an example `foo.bar.baz` -> ('foo.bar', 'baz')`. Also worth specifying the edge case for a root qualname

File: torch/fx/experimental/fuser.py [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+            return False
+        if type(modules[current_node.target]) is not pattern_node:
+            return False
+    return True
+
+
+def fuse(model):
+    patterns = [(torch.nn.Conv2d, torch.nn.BatchNorm2d), (torch.nn.Conv1d, torch.nn.BatchNorm1d), (torch.nn.Conv3d, torch.nn.BatchNorm3d)]
+    for pattern in patterns:
+        for node in model.graph.nodes:
```
Comment (jamesr66a) [inline torch/fx/experimental/fuser.py:35]: Looping over (# patterns) * (# nodes) seems a bit excessive. Is it possible to formulate the patterns as an FSM or something to only iterate once?

File: torch/fx/experimental/fuser.py [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+            modules = dict(model.named_modules())
+            if matches_module_pattern(pattern, node, modules):
+                if len(node.args[0].users) > 1:  # Output of conv is used by other nodes
+                    continue
+                fused_conv = fuse_conv_bn_eval(modules[node.args[0].target], modules[node.target])
+                parent_name, name = _parent_name(node.args[0].target)
+                setattr(dict(model.named_modules())[parent_name], name, fused_conv)
```
Comment (jamesr66a) [inline torch/fx/experimental/fuser.py:42]: Instead of mutating Modules, can we simply construct a new dict of qualname -> Module and pass that into the `GraphModule` constructor? https://github.com/pytorch/pytorch/blob/cdc2d2843b2bb7ef1f7192718a24bfabb93b3d36/torch/fx/graph_module.py#L152

Line 36 can be hoisted out of the loop nest here and modified when we find a match. Also, we should make a new copy of the Graph. Then we can just pass that dict and the new graph to the GraphModule constructor

File: test/test_fx_experimental.py [hunk 11]
```diff
@@ -227,6 +235,20 @@ def get_node_to_latency_mapping(fx_module: GraphModule):
...
+        fused = fuse(traced)
+
+        self.assertTrue(all(not isinstance(m, torch.nn.BatchNorm2d) for m in fused.modules()))
+
+        N, C, H, W = 20, 3, 224, 224
+        inp = torch.randn(N, C, H, W)
+
+        self.assertEqual(fused(inp), rn18(inp))
+
+
```
Comment (jamesr66a) [inline test/test_fx_experimental.py:251]: nit: extra blank line

**Review by jamesr66a — COMMENTED** [review]
Can you type annotate the functions in this PR? There are some issues in the code that would be revealed by mypy if the functions were typed

File: torch/fx/experimental/fuser.py (hunk 9 above)
Comment (Chillee) [reply torch/fx/experimental/fuser.py:35]: well ... yes. I just didn't think it was worth it for now. We only have 3 patterns so it's a relatively small constant factor increase.

File: torch/fx/experimental/fuser.py (hunk 7 above)
Comment (Chillee) [reply torch/fx/experimental/fuser.py:3]: I'll do the "incorrect" way for now :P 

If that's the correct way we can do it in another PR.

Comment (Chillee) [reply torch/fx/experimental/fuser.py:3]: Actually, that doesn't fix it lol.

File: torch/fx/experimental/fuser.py (hunk 8 above)
Comment (Chillee) [reply torch/fx/experimental/fuser.py:7]: Shoulda been a reviewer for the initial code :P https://github.com/Chillee/pytorch/blob/fusion/test/fx/quantization.py#L145

File: torch/fx/experimental/fuser.py (hunk 9 above)
Comment (jamesr66a) [reply torch/fx/experimental/fuser.py:37]: Wait till we have many patterns, then nobody is gonna want to touch it :p

File: torch/fx/experimental/fuser.py (hunk 10 above)
Comment (Chillee) [reply torch/fx/experimental/fuser.py]: Kind of resolved? There are issues with just having a single dict, so the quantized fuser does something similar.

File: torch/fx/experimental/fuser.py [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+        for node in new_graph.nodes:
+            if matches_module_pattern(pattern, node, modules):
+                if len(node.args[0].users) > 1:  # Output of conv is used by other nodes
+                    continue
+                fused_conv = fuse_conv_bn_eval(modules[node.args[0].target], modules[node.target])
+                parent_name, name = _parent_name(node.args[0].target)
+                setattr(modules[parent_name], name, fused_conv)
+                node.replace_all_uses_with(node.args[0])
+                new_graph.erase_node(node)
+    return fx.GraphModule(dict(model.named_modules()), new_graph)
```
Comment (jamesr66a) [inline torch/fx/experimental/fuser.py:49]: Is this right? above we're mutating the `modules` object, but then passing in a new dict created from the original symbolically traced Module to the constructor. Does this work?

File: torch/fx/experimental/fuser.py [hunk 13]
```diff
@@ -0,0 +1,49 @@
...
+        if not isinstance(current_node.target, str):
+            return False
+        if current_node.target not in modules:
+            return False
+        if type(modules[current_node.target]) is not expected_type:
+            return False
+    return True
+
+
+def fuse(model: torch.nn.Module) -> torch.nn.Module:
```
Comment (jamesr66a) [inline torch/fx/experimental/fuser.py:34]: Probably want a docstring here too since this is the user-visible API

File: torch/fx/experimental/fuser.py (hunk 12 above)
Comment (Chillee) [reply torch/fx/experimental/fuser.py:49]: So the way this works is that the modules object is just a pointer to the underlying module of the model - we don't actually care what's in the modules object itself. I was a bit confused by this (and thought that perhaps both this code and the quantization fuser were doing something wrong), but I think it's correct now.

So imagine you had a module with just a `conv1` in it, then `named_modules` would return something like
```
{
   '': Module(conv1: Conv2d(...))
   "conv1": Conv2d(...)
}
```

Basically, the reason the code looks for the parent and then indexes with the qualified name is that this provides a pointer to the actual module in the model, and not that we're modifying the module class itself.

**Review by jamesr66a — APPROVED** [review]

**Comment by ilia-cher** [comment]
is this PR related to pytorch_linux_xenial_py3_6_gcc5_4_test breakage?
https://app.circleci.com/pipelines/github/pytorch/pytorch/242084/workflows/500a3a4c-23d5-4272-9616-8a0850178502/jobs/9110657
https://ezyang.github.io/pytorch-ci-hud/build/pytorch-master

**Comment by Chillee** [comment]
Yeah it is -what's the easiest way of fixing it?

The fix is to add an assert on line 35 of that file

`assert(isinstance(node.target, str))`
